### ISSUE TITLE
Unsound abstraction of Dataset::try_from

### ISSUE THREAD

**Issue by lizhuohua** [spglib/spglib-rs#1]
Hi. I am developing a static analysis tool for Rust and it detected a potential vulnerability in the `TryFrom` trait implementation for `Dataset`:

https://github.com/spglib/spglib-rs/blob/37ceb33ea13705aa1cfaaa0636a5d40cfa4547a2/src/dataset.rs#L69

First the argument `value` is dereferenced to get `ptr`, an object of `SpglibDataset`. Then the ownership of `ptr` is transferred to several vectors using `Vec::from_raw_parts`. For example:

https://github.com/spglib/spglib-rs/blob/37ceb33ea13705aa1cfaaa0636a5d40cfa4547a2/src/dataset.rs#L95

Finally, these vectors are returned.

The problem occurs when users try to invoke `Dataset::try_from` twice on the same `*mut ffi::SpglibDataset`. This would create two aliases to the same object, which violates Rust's ownership system. The example code that triggers the problem:

```rust
use spglib::dataset::Dataset;
use spglib_sys::spg_get_dataset;
use std::convert::TryFrom;

fn main() {
    let mut lattice = [[4.0, 0.0, 0.0], [0.0, 4.0, 0.0], [0.0, 0.0, 3.0]];
    // This unsafe is OK, it calls an external C API to construct a dataset
    let spglib_dataset_ptr =
        unsafe { spg_get_dataset(&mut lattice[0], &mut [0.0, 0.0, 0.0], &1, 1, 0.00001) };

    // This would cause a double-free, because `try_from` gets ownership from input
    let dataset1 = Dataset::try_from(spglib_dataset_ptr);
    let dataset2 = Dataset::try_from(spglib_dataset_ptr);
}
```

Look forward to you reply! Thank you very much.

**Comment by seatonullberg** [comment]
Hi, thanks for taking the time to investigate this! The code sample you've provided is indeed problematic, however, it does not represent the typical use case of `Dataset::try_from`. Datasets should be constructed using `Dataset::new` as shown below:

```rust
use spglib::cell::Cell;
use spglib::dataset::Dataset;

let lattice = [[4., 0., 0.], [0., 4., 0.], [0., 0., 4.]];
let positions = [[0., 0., 0.], [0.5, 0.5, 0.5]];
let types = [1, 1];
let mut bcc_cell = Cell::new(&lattice, &positions, &types);
let dataset = Dataset::new(&mut bcc_cell, 1e-5);
assert_eq!(dataset.hall_number, 529);
```

This constructor creates a new `*mut ffi::SpglibDataset` on each call, thus preventing a double-free. The implementation of `TryFrom` was a convenience for me to use internally in this constructor and is not intended to be exposed in the user-facing API.  I agree that it should be impossible for users to violate Rust's ownership system in this way and would be interested in any suggestions you might have to prevent the `TryFrom` implementation from being used outside of the `spglib` crate.

**Comment by lizhuohua** [comment]
Thanks for your reply! That makes sense.

I don't think we can control the visibility of a trait implementation. Since both `TryFrom` and `Dataset` are public, `Dataset::try_from` will be public as well.

Probably it is better to define a wrapper struct that encapsulates raw pointers, so that `Dataset::try_from` takes this wrapper struct as argument instead of the raw pointer. This way, the ownership can be correctly handled by Rust.

**Comment by seatonullberg** [comment]
Good idea, that sounds like a simple fix. I will work on a PR for this soon and merge it into the next release. These bindings are currently 1 minor release behind `spglib` anyways so it would be a good opportunity to synchronize the versions.

**Comment by seatonullberg** [comment]
Hi @lizhuohua, I've introduced a fix for this in #2. I believe it works as we discussed, but I will wait for your review to merge. Thanks again for identifying this bug.

**Comment by lizhuohua** [comment]
Looks good to me.

Thanks for your work!
